### PR TITLE
Add an option to specify the max response size cached

### PR DESCRIPTION
**Pull request by JunTaoLuo**
cc @Tratcher @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using Microsoft.AspNetCore.ResponseCaching.Internal;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    public class ResponseCachingOptions
+    {
+        /// <summary>
+        /// The largest cacheable size for the response body in bytes.
+        /// </summary>
+        public int? MaximumCachedBodySize { get; set; }
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs]: Should we specify a default size? Also could consider the naming `MaximumCacheableBodySize` or `MaximumBodySizeToCache`.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs]: `long`, not `int?`. Default to 1mb for starters?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 2]
```diff
@@ -598,7 +594,9 @@ internal void FinalizeCachingHeaders()
 
         internal void FinalizeCachingBody()
         {
-            if (CacheResponse && ResponseCacheStream.BufferingEnabled)
+            if (CacheResponse &&
+                ResponseCacheStream.BufferingEnabled &&
+                !(_options.MaximumCachedBodySize < ResponseCacheStream.BufferedStream.Length))
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: BufferingEnabled should be false as soon as MaximumCachedBodySize is exceeded

**Comment by JunTaoLuo** [comment]
:up::date:

**Comment by Tratcher** [comment]
FYI: It looks like MemoryStream.Dispose doesn't actually free the memory. You need to call SetLength and Capacity to shrink it.

**Comment by Tratcher** [comment]
:shipit:

**Pull request merged**